packages/cinterop/src/darwin/kotlin/io/realm/interop/RealmInterop.kt

```diff
@@ -883,6 +892,138 @@ actual object RealmInterop {
         )
     }
 
+    // TODO sync config shouldn't be null
+    actual fun realm_app_new(appConfig: NativePointer, basePath: String): NativePointer {
+        val syncClientConfig = realm_wrapper.realm_sync_client_config_new()
+        realm_wrapper.realm_sync_client_config_set_base_file_path(syncClientConfig, basePath)
+
+        // TODO add metadata mode to config
+        realm_wrapper.realm_sync_client_config_set_metadata_mode(syncClientConfig, realm_wrapper.realm_sync_client_metadata_mode_e.RLM_SYNC_CLIENT_METADATA_MODE_DISABLED)
+        return CPointerWrapper(realm_wrapper.realm_app_new(appConfig.cptr(), syncClientConfig))
+    }
+
+    actual fun realm_app_log_in_with_credentials(app: NativePointer, credentials: NativePointer, callback: CinteropCallback) {
+        realm_wrapper.realm_app_log_in_with_credentials(
+            app.cptr(),
+            credentials.cptr(),
+            staticCFunction { userdata, user, error ->
+                val callback = safeUserData<CinteropCallback>(userdata)
+                if (error == null) {
+                    callback.onSuccess(CPointerWrapper(user))
+                } else {
+                    callback.onError(toKException(error.pointed))
+                }
+            },
+            StableRef.create(callback).asCPointer(),
+            staticCFunction { userdata -> disposeUserData<Callback>(userdata) }
+        )
+    }
+
+    private val newRequestLambda = staticCFunction<COpaquePointer?,
+        CValue<realm_http_request_t>,
+        COpaquePointer?,
+        realm_http_completion_func_t?,
+        Unit>
+    { userdata, request, completionData, callback ->
+        safeUserData<NetworkTransport>(userdata).let { networkTransport ->
+            request.useContents {
+                val headerMap = mutableMapOf<String, String>()
+                for (i in 0 until num_headers.toInt()) {
```

Comment:
Where is `num_headers` defined? 🤔

Reply:
The `CValue<T>.useContents(block: T.() -> R): R` promotes the `request` as the receiver, so all referencess like `num_headers`, `headers`, `method`, etc. are from the `request`. I have tried to change some of the `apply`s to `let` and name the target explicitly, but it is not possible with `useContents`. Only option would be to prepend `this`, but I didn't find it more explanatory.

Reply:
Ah okay... that's fine. One of the problems with too much Kotlin magic in Github reviews I guess :D